snap/validate.go

```diff
@@ -263,6 +264,44 @@ func validateAppSocket(socket *SocketInfo) error {
 	return validateSocketAddr(socket, "listen-stream", socket.ListenStream)
 }
 
+type appStartupOrder int
+
+const (
+	orderBefore appStartupOrder = iota
+	orderAfter
+)
+
+func validateAppStartup(app *AppInfo, order appStartupOrder) error {
+	orderStr := "before"
+	deps := app.Before
+	if order == orderAfter {
+		orderStr = "after"
+		deps = app.After
+	}
+
+	for _, dep := range deps {
+		// dependency is not defined
+		if other, ok := app.Snap.Apps[dep]; !ok {
+			return fmt.Errorf("cannot validate app %q startup ordering: needs to be started %s %q, but %q is not defined",
+				app.Name, orderStr, dep, dep)
+		} else {
```

Comment:
(nitpick) I would prefer to avoid the "else" here. Because we return above (c.f. effective go:  `In the Go libraries, you'll find that when an if statement doesn't flow into the next statement—that is, the body ends in break, continue, goto, or return—the unnecessary else is omitted. `)